SourceCode/Scene/CRTMesh.cpp

```diff
@@ -1,19 +1,145 @@
 #include "CRTMesh.h"
 
-CRTMesh::CRTMesh(const std::vector<CRTVector>& vertices, const std::vector<int>& triangleVertIndices) : 
-    vertices(vertices), triangleVertIndices(triangleVertIndices)
-{}
+void CRTMesh::calculateFaceNormals() {
+    for (int i = 0; i < triangleVertIndices.size(); i += VERTICES) {
+        CRTVector v0 = vertices[triangleVertIndices[i + 0]];
+        CRTVector v1 = vertices[triangleVertIndices[i + 1]];
+        CRTVector v2 = vertices[triangleVertIndices[i + 2]];
+        CRTVector V1 = v1 - v0;
+        CRTVector V2 = v2 - v0;
 
-std::vector<CRTTriangle> CRTMesh::getTriangles() const
+        CRTVector normal = cross(V1, V2); // we set it to the counterclockwise cross-product
+        normal.normalize();
+        faceNormals[i / VERTICES] = std::move(normal);
+    }
+}
+
+void CRTMesh::calculateVertexNormals() {
+    for (int i = 0; i < vertices.size(); i++) {
+        CRTVector vNormal;
+        int triangleParticipations = 0;
+        for (int j = 0; j < triangleVertIndices.size(); j++) {
```

Comment:
This could be a serious bottleneck, consider optimizing this calculation while you parse the scene.